File: frontend/src/lib/api/supplier-assessment.ts
import apiClient from "./client";

export interface AssessmentTemplate {
  id: string;
  organization_id: string;
  title: string;
  description: string;
  is_default: boolean;
  created_by: string;
  created_at: string;
  question_count: number;
}

export interface AssessmentQuestion {
  id: string;
  section: string;
  question_text: string;
  question_type: string;
  options: string[];
  csddd_article: string;
  weight: number;
  is_required: boolean;
  sort_order: number;
}

export interface SupplierAssessment {
  id: string;
  organization_id: string;
  template_id: string;
  supplier_id: string;
  status: string;
  reference_code: string;
  token_expires_at: string;
  created_at: string;
  submitted_at: string | null;
  portal_link?: string;
}

export interface GapItem {
  question_id: string;
  section: string;
  csddd_article: string;
  question_text: string;
  answer_given: string;
  expected_answer: string;
  severity: string;
  recommendation: string;
}

export interface SectionScore {
  section: string;
  total_questions: number;
  answered: number;
  gaps: number;
  traffic_light: string;
}

export interface GapReport {
  assessment_id: string;
  supplier_id: string;
  section_scores: SectionScore[];
  gaps: GapItem[];
  overall_traffic_light: string;
  total_gaps: number;
  critical_gaps: number;
  generated_at: string;
}

// Portal types (public)
export interface PortalQuestion extends AssessmentQuestion {
  saved_answer: string;
}

export interface PortalAssessment {
  assessment_id: string;
  template_title: string;
  status: string;
  reference_code: string;
  expires_at: string;
  questions: PortalQuestion[];
}

const BASE = "/supplier-assessments";
const PUBLIC = "/supplier-portal";

export async function seedDefaultTemplate(): Promise<AssessmentTemplate> {
  const { data } = await apiClient.post(`${BASE}/templates/seed`);
  return data;
}

export async function listTemplates(): Promise<AssessmentTemplate[]> {
  const { data } = await apiClient.get(`${BASE}/templates`);
  return data;
}

export async function getTemplate(id: string): Promise<AssessmentTemplate & { questions: AssessmentQuestion[] }> {
  const { data } = await apiClient.get(`${BASE}/templates/${id}`);
  return data;
}

export async function createAssessment(templateId: string, supplierId: string): Promise<SupplierAssessment> {
  const { data } = await apiClient.post(`${BASE}/`, { template_id: templateId, supplier_id: supplierId });
  return data;
}

export async function listAssessments(status?: string): Promise<SupplierAssessment[]> {
  const { data } = await apiClient.get(`${BASE}/`, { params: status ? { status } : {} });
  return data;
}

export async function getAssessment(id: string): Promise<SupplierAssessment> {
  const { data } = await apiClient.get(`${BASE}/${id}`);
  return data;
}

export async function getGapReport(id: string): Promise<GapReport> {
  const { data } = await apiClient.get(`${BASE}/${id}/gap-report`);
  return data;
}

// Public portal functions (no auth required — called directly by browser)
export async function portalGetAssessment(token: string): Promise<PortalAssessment> {
  const { data } = await apiClient.get(`${PUBLIC}/assessment/${token}`);
  return data;
}

export async function portalSaveProgress(token: string, answers: Record<string, string>): Promise<void> {
  await apiClient.post(`${PUBLIC}/assessment/${token}/save`, { answers });
}

export async function portalSubmit(token: string, answers: Record<string, string>, email: string): Promise<{ reference_code: string }> {
  const { data } = await apiClient.post(`${PUBLIC}/assessment/${token}/submit`, {
    answers,
    submitter_email: email,
    confirm_accuracy: true,
  });
  return data;
}

File: backend/domain/supplier_assessment.py
"""Domain model — Supplier Self-Assessment CSDDD (Art. 10 Abs. 2 lit. a).

Art. 10 Abs. 2 lit. a CSDDD allows using self-assessment questionnaires as a
prevention measure for indirect business partners (Tier 2+).

Security constraints:
  - submitted_by_email NEVER returned in API responses
  - IP address NEVER returned in API responses
  - Token = JWT with assessment_id + supplier_id + expires_at
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime


@dataclass
class AssessmentQuestion:
    """A single question in an assessment template."""

    id: str
    template_id: str | None  # None for seed/library questions
    section: str  # AssessmentSection
    question_text: str
    question_type: str  # QuestionType
    options: list[str]  # choices for multiple_choice
    csddd_article: str  # e.g. "Art. 10 Annex I"
    weight: int  # gap severity weight 1-5
    is_required: bool
    sort_order: int
    is_active: bool


@dataclass
class AssessmentTemplate:
    """A reusable questionnaire template owned by an organization."""

    id: str
    organization_id: str
    title: str
    description: str
    is_default: bool  # True for the system-seeded CSDDD template
    created_by: str
    created_at: datetime
    updated_at: datetime
    question_count: int = 0


@dataclass
class SupplierAssessment:
    """An instance of a template sent to a specific supplier.

    submitted_by_email is stored internally but NEVER returned in API responses.
    """

    id: str
    organization_id: str
    template_id: str
    supplier_id: str
    token_hash: str  # SHA-256 of the JWT; raw token never stored
    token_expires_at: datetime
    status: str  # AssessmentStatus
    reference_code: str  # shown to supplier after submission
    created_at: datetime
    updated_at: datetime
    submitted_at: datetime | None


@dataclass
class AssessmentResponse:
    """A single answer given by the supplier."""

    id: str
    assessment_id: str
    question_id: str
    answer_value: str  # JSON-encoded for complex types
    answered_at: datetime


@dataclass
class GapItem:
    question_id: str
    section: str
    csddd_article: str
    question_text: str
    answer_given: str
    expected_answer: str
    severity: str  # GapSeverity
    recommendation: str


@dataclass
class SectionScore:
    section: str
    total_questions: int
    answered: int
    gaps: int
    traffic_light: str  # TrafficLight


@dataclass
class GapReport:
    assessment_id: str
    supplier_id: str
    section_scores: list[SectionScore]
    gaps: list[GapItem]
    overall_traffic_light: str  # TrafficLight
    total_gaps: int
    critical_gaps: int
    generated_at: datetime
